device/cuda/src/seeding2/seed_finding.cu

```diff
@@ -0,0 +1,108 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <memory>
+#include <traccc/cuda/seeding2/kernels/kd_tree_kernel.hpp>
+#include <traccc/cuda/seeding2/kernels/seed_finding_kernel.hpp>
+#include <traccc/cuda/seeding2/kernels/write_output_kernel.hpp>
+#include <traccc/cuda/seeding2/seed_finding.hpp>
+#include <traccc/cuda/seeding2/types/internal_seed.hpp>
+#include <traccc/cuda/seeding2/types/kd_tree.hpp>
+#include <traccc/cuda/utils/definitions.hpp>
+#include <traccc/edm/seed.hpp>
+#include <traccc/edm/spacepoint.hpp>
+#include <traccc/seeding/detail/lin_circle.hpp>
+#include <traccc/seeding/detail/seeding_config.hpp>
+#include <vecmem/memory/memory_resource.hpp>
+#include <vecmem/memory/unique_ptr.hpp>
+#include <vector>
+
+namespace traccc::cuda {
+std::tuple<vecmem::unique_alloc_ptr<internal_spacepoint<spacepoint>[]>,
+           uint32_t>
+copy_sps_to_device(vecmem::memory_resource& mr,
+                   const spacepoint_container_types::const_view& sps) {
+    std::chrono::high_resolution_clock::time_point time_ex_start =
+        std::chrono::high_resolution_clock::now();
+
+    const traccc::spacepoint_container_types::const_device p(sps);
+
+    std::vector<internal_spacepoint<spacepoint>> spacepoints;
+
+    for (std::size_t i = 0; i < p.get_items().size(); ++i) {
+        for (std::size_t j = 0; j < p.get_items().at(i).size(); ++j) {
+            const traccc::spacepoint& sp = p.get_items().at(i).at(j);
+
+            spacepoints.push_back(
+                internal_spacepoint<spacepoint>(p, {i, j}, {0.f, 0.f}));
+        }
+    }
```

Comment:
In principle yes, but on the other hand this is just data pre-processing that would not really exist in any sort of production code, so it's not performance-relevant per sé.

Reply:
I don't understand. :confused: The container structures that we use in this project did not come out of thin air. We have the container layout for spacepoints that we have because in ATLAS we use such a layout. So such conversions may very well need to stay with us in the long term.

Reply:
I agree that these data conversions make some sense when moving data from the host to the device, but obviously such movements would be minimized in a production setting, and it makes little sense for data produced by another GPU algorithm to be in such a format. So I would consider this more of an edge case than something that would be executed with non-negligible frequency. :confused: